File: frameworks/java/javalin/UnitTests/src/test/java/unittests/ItemsTests.java
package unittests;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.Map;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.javalin.testtools.JavalinTest;
import io.javalin.testtools.Response;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

class ItemsTests extends JavalinApp {

    // rb:test items.read
    @Test
    @Tag("items.read")
    void aRowIsReadByTheIdInThePath() {
        JavalinTest.test(app(), (server, client) -> Answer.is(Expected.row(17), client.get("/items/17")));
    }

    // rb:test items.head
    @Test
    @Tag("items.head")
    void headRunsTheGetHandlerAndSendsNoBody() {
        JavalinTest.test(app(), (server, client) -> {
            HttpResponse<byte[]> response = send(request(client, "/items/17").method("HEAD", HttpRequest.BodyPublishers.noBody()));

            assertEquals(200, response.statusCode());
            assertTrue(Answer.header(response, "content-type").startsWith("application/json"));
            assertEquals(0, response.body().length);
        });
    }

    // rb:test items.create
    @Test
    @Tag("items.create")
    void aCreatedItemIsTheRowAfterTheLast() {
        JavalinTest.test(app(), (server, client) -> {
            Response response = client.post("/items", Expected.text("items.new.json"));

            assertEquals(201, response.code());
            assertEquals("/items/1426", Answer.header(response, "location"));
            Answer.is(numbered(1426), response);
        });
    }

    // rb:test items.replace
    @Test
    @Tag("items.replace")
    void aReplacedItemTakesTheIdInThePath() {
        JavalinTest.test(app(), (server, client) -> Answer.is(numbered(17), client.put("/items/17", Expected.text("items.new.json"))));
    }

    // rb:test items.update
    @Test
    @Tag("items.update")
    void aPatchIsMergedOntoTheRow() {
        ObjectNode merged = Expected.row(17);
        for (Map.Entry<String, JsonNode> field : Expected.json("items.patch.json").properties()) {
            merged.set(field.getKey(), field.getValue());
        }

        JavalinTest.test(app(), (server, client) -> Answer.is(merged, client.patch("/items/17", Expected.text("items.patch.json"))));
    }

    // rb:test items.delete
    @Test
    @Tag("items.delete")
    void aDeleteIsAnswered204WithNoBody() {
        JavalinTest.test(app(), (server, client) -> {
            Response response = client.delete("/items/17");

            assertEquals(204, response.code());
            assertEquals("", response.body().string());
        });
    }

    private static ObjectNode numbered(int id) {
        ObjectNode item = Expected.JSON.createObjectNode().put("id", id);
        item.setAll((ObjectNode) Expected.json("items.new.json"));
        return item;
    }
}

File: frameworks/dotnet/hardened/UnitTests/ItemsTests.cs
namespace UnitTests;

public class ItemsTests
{
    private static JsonNode Row(int id) => Expected.Json("items.large.json")["items"]![id - 1]!;

    // rb:test items.read
    [ModuleTest]
    [Trait("corpus", "items.read")]
    public async Task A_row_is_read_by_its_id(ITestWebApp app)
    {
        TestWebResponse response = await app.Get("/items/17");

        response.Assert.Ok();
        await Answer.Is(Row(17), response);
    }

    // rb:test items.head
    [ModuleTest]
    [Trait("corpus", "items.head")]
    public async Task Head_answers_the_headers_alone(ITestWebApp app)
    {
        TestWebResponse response = await app.Request("HEAD", null, "/items/17");

        response.Assert.Ok();
        Assert.StartsWith("application/json", Answer.Header(response, "Content-Type"));
        Assert.Equal(0, response.Body.Length);
    }

    // rb:test items.create
    [ModuleTest]
    [Trait("corpus", "items.create")]
    public async Task Create_answers_the_next_id_and_its_location(ITestWebApp app)
    {
        JsonNode item = Expected.Json("items.new.json");
        int next = (int)Expected.Json("items.large.json")["count"]! + 1;

        TestWebResponse response = await app.Request("POST", null, "/items", request => request.RawBody(item.ToJsonString()));

        Assert.Equal(201, response.StatusCode);
        Assert.Equal($"/items/{next}", Answer.Header(response, "Location"));
        JsonObject created = item.DeepClone().AsObject();
        created.Insert(0, "id", next);
        await Answer.Is(created, response);
    }

    // rb:test items.replace
    [ModuleTest]
    [Trait("corpus", "items.replace")]
    public async Task Replace_answers_the_body_under_the_path_id(ITestWebApp app)
    {
        JsonNode item = Expected.Json("items.new.json");

        TestWebResponse response = await app.Request("PUT", null, "/items/17", request => request.RawBody(item.ToJsonString()));

        response.Assert.Ok();
        JsonObject replaced = item.DeepClone().AsObject();
        replaced.Insert(0, "id", 17);
        await Answer.Is(replaced, response);
    }

    // rb:test items.update
    [ModuleTest]
    [Trait("corpus", "items.update")]
    public async Task Update_answers_the_row_with_the_patch_applied(ITestWebApp app)
    {
        TestWebResponse response = await app.Request("PATCH", null, "/items/17", request => request.RawBody("""{"priceCents":1999,"inStock":false}"""));

        response.Assert.Ok();
        JsonObject patched = Row(17).DeepClone().AsObject();
        patched["priceCents"] = 1999;
        patched["inStock"] = false;
        await Answer.Is(patched, response);
    }

    // rb:test items.delete
    [ModuleTest]
    [Trait("corpus", "items.delete")]
    public async Task Delete_answers_no_content(ITestWebApp app)
    {
        TestWebResponse response = await app.Delete("/items/17");

        Assert.Equal(204, response.StatusCode);
        Assert.Equal(0, response.Body.Length);
    }

    [ModuleTest]
    public async Task A_write_to_a_row_that_is_not_there_is_not_found(ITestWebApp app)
    {
        (await app.Delete("/items/999999")).Assert.NotFound();
    }
}
